public/app/core/components/TimelineChart/timeline.test.ts

```diff
@@ -0,0 +1,211 @@
+    const orientCallback = jest.mocked(uPlot.orient).mock.calls[jest.mocked(uPlot.orient).mock.calls.length - 1][2];
+    const methods = {
+      moveTo: jest.fn(() => {}),
+      lineTo: jest.fn(() => {}),
+      rect: jest.fn(() => {}),
+      arc: jest.fn(() => {}),
+      bezierCurveTo: jest.fn(() => {}),
+    };
+
+    orientCallback(
+      mockUplot.series[0],
+      mockUplot.data[0] as number[],
+      mockUplot.data[1] as number[],
+      mockUplot.scales.x,
+      mockUplot.scales.y,
+      jest.fn(() => 1),
+      jest.fn(() => 1),
+      0,
+      0,
+      100,
+      100,
+      methods.moveTo,
+      methods.lineTo,
+      methods.rect,
+      methods.arc,
+      methods.bezierCurveTo
+    );
+
+    return methods;
+  };
+
+  describe('#drawPoints', () => {
+    it('returns a `drawPoints` method when a `formatValue` function is provided', () => {
+      const config = getConfig(buildTestCoreOptions({ formatValue: () => 'foo' }));
+      expect(typeof config.drawPoints).toBe('function');
+    });
+
+    it('returns false for `drawPoints` when no `formatValue` function is provided', () => {
+      const config = getConfig(buildTestCoreOptions());
+      expect(config.drawPoints).toBe(false);
+    });
+
+    it('returns false for `drawPoints` if the visibility mode is `never`', () => {
+      const config = getConfig(buildTestCoreOptions({ formatValue: () => 'foo', showValue: VisibilityMode.Never }));
+      expect(config.drawPoints).toBe(false);
+    });
+
+    it('returns a function for `drawPoints` if the conditions are met', () => {
+      const { drawPoints } = getConfig(buildTestCoreOptions({ formatValue: () => 'foo' }));
+      if (!drawPoints) {
+        throw new Error('drawPoints is not defined');
+      }
+      const mockUplot = buildMockUplotInstance();
+      expect(drawPoints(mockUplot, 1, 2, 3, null)).toBe(false);
+      expect(uPlot.orient).toHaveBeenCalledWith(mockUplot, 1, expect.any(Function));
+    });
+
+    describe('#drawPaths', () => {
+      describe('null and NaN values', () => {
+        // these tests are attempting to determine whether `shouldDrawYVal` is returning false
+        // and preventing a draw for a given value. this is being done by checking the number of
+        // calls to `rect` by the `orient` callback created by a `drawPaths` call. Even
+
+        it('should draw boxes for null values when hasMappedNull returns true and isDiscrete returns true', () => {
+          const { drawClear, drawPaths } = getConfig(
+            buildTestCoreOptions({
+              hasMappedNull: jest.fn(() => true),
+              isDiscrete: jest.fn(() => true),
+              formatValue: () => 'foo',
+            })
+          );
+          const mockUplot = buildMockUplotInstance([[0], [null]]);
+
+          drawClear(mockUplot);
+          drawPaths(mockUplot, 1, 0, 1);
+
+          const { rect } = callOrientCallback(mockUplot);
+          expect(rect).toHaveBeenCalledTimes(2);
```

Comment:
if we can think of a better proxy for this, I'd be happy to explore it, this number-of-calls check feels brittle.